src/repr/src/scalar.rs

```diff
@@ -2241,6 +2241,307 @@ impl<'a> ScalarType {
             (s, o) => ScalarBaseType::from(s) == ScalarBaseType::from(o),
         }
     }
+
+    /// Returns various interesting datums for a ScalarType (max, min, 0 values,
+    /// etc.).
+    pub fn interesting_datums(&self) -> impl Iterator<Item = Datum> {
+        // TODO: Is there a better way than packing everything into Lazys and
```

Comment:
I don't know of a better way to handle this.

As an aside, you could have many fewer characters to instantiate these things using a relatively simple macro:

```diff
diff --git a/src/repr/src/scalar.rs b/src/repr/src/scalar.rs
index 4bcdcd3a5..9afc846d9 100644
--- a/src/repr/src/scalar.rs
+++ b/src/repr/src/scalar.rs
@@ -2249,36 +2249,45 @@ impl<'a> ScalarType {
         // Rows? `&[Datum::X(x)]` doesn't seem to work because some Datum
         // variants need to call non-const functions to construct themselves.
 
+        macro_rules! interesting {
+            (Float32 ($($val:expr,)*)) => {
+                interesting!(Float32 {
+                    $(OrderedFloat($val)),*,
+                })
+            };
+            ($datum_type:ident {$($val:expr,)*}) => {
+                Lazy::new(|| {
+                    Row::pack_slice(&[
+                        $(Datum::$datum_type($val)),*,
+                    ])
+                })
+            };
+        }
+
         static BOOL: Lazy<Row> = Lazy::new(|| Row::pack_slice(&[Datum::True, Datum::False]));
-        static INT16: Lazy<Row> = Lazy::new(|| {
-            Row::pack_slice(&[
-                Datum::Int16(0),
-                Datum::Int16(1),
-                Datum::Int16(-1),
-                Datum::Int16(i16::MIN),
-                Datum::Int16(i16::MIN + 1),
-                Datum::Int16(i16::MAX),
-            ])
+        static INT16: Lazy<Row> = interesting!(Int16 {
+            0,
+            1,
+            -1,
+            i16::MIN,
+            i16::MIN + 1,
+            i16::MAX,
         });
-        static INT32: Lazy<Row> = Lazy::new(|| {
-            Row::pack_slice(&[
-                Datum::Int32(0),
-                Datum::Int32(1),
-                Datum::Int32(-1),
-                Datum::Int32(i32::MIN),
-                Datum::Int32(i32::MIN + 1),
-                Datum::Int32(i32::MAX),
-            ])
+        static INT32: Lazy<Row> = interesting!(Int32 {
+            0,
+            1,
+            -1,
+            i32::MIN,
+            i32::MIN + 1,
+            i32::MAX,
         });
-        static INT64: Lazy<Row> = Lazy::new(|| {
-            Row::pack_slice(&[
-                Datum::Int64(0),
-                Datum::Int64(1),
-                Datum::Int64(-1),
-                Datum::Int64(i64::MIN),
-                Datum::Int64(i64::MIN + 1),
-                Datum::Int64(i64::MAX),
-            ])
+        static INT64: Lazy<Row> = interesting!(Int64 {
+            0,
+            1,
+            -1,
+            i64::MIN,
+            i64::MIN + 1,
+            i64::MAX,
         });
         static UINT16: Lazy<Row> = Lazy::new(|| {
             Row::pack_slice(&[Datum::UInt16(0), Datum::UInt16(1), Datum::UInt16(u16::MAX)])
@@ -2289,20 +2298,19 @@ impl<'a> ScalarType {
         static UINT64: Lazy<Row> = Lazy::new(|| {
             Row::pack_slice(&[Datum::UInt64(0), Datum::UInt64(1), Datum::UInt64(u64::MAX)])
         });
-        static FLOAT32: Lazy<Row> = Lazy::new(|| {
-            Row::pack_slice(&[
-                Datum::Float32(OrderedFloat(0.0)),
-                Datum::Float32(OrderedFloat(1.0)),
-                Datum::Float32(OrderedFloat(-1.0)),
-                Datum::Float32(OrderedFloat(f32::MIN)),
-                Datum::Float32(OrderedFloat(f32::MIN_POSITIVE)),
-                Datum::Float32(OrderedFloat(f32::MAX)),
-                Datum::Float32(OrderedFloat(f32::EPSILON)),
-                Datum::Float32(OrderedFloat(f32::NAN)),
-                Datum::Float32(OrderedFloat(f32::INFINITY)),
-                Datum::Float32(OrderedFloat(f32::NEG_INFINITY)),
-            ])
-        });
+        static FLOAT32: Lazy<Row> = interesting!(Float32(
+            0.0,
+            1.0,
+            -1.0,
+            f32::MIN,
+            f32::MIN_POSITIVE,
+            f32::MAX,
+            f32::EPSILON,
+            f32::NAN,
+            f32::INFINITY,
+            f32::NEG_INFINITY,
+        ));
+
         static FLOAT64: Lazy<Row> = Lazy::new(|| {
             Row::pack_slice(&[
                 Datum::Float64(OrderedFloat(0.0)),
```

Reply:
If this appeals to you at all, I'd be glad to do the rest of the typing for you, or you can disregard.

Reply:
Hmm I'm ambivalent about this, so I think I'll merge as is. If you or anyone has opinions we can do that in followup.